infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationManager.java

```diff
@@ -33,32 +34,100 @@ public class CorfuReplicationManager {
         this.upgradeManager = upgradeManager;
+        this.topology = topology;
+        this.replicationContext = replicationContext;
+        this.replicationSessionToRuntimeParams = new HashMap<>();
     }
 
     /**
-     * Start log replication by instantiating a runtime for each session
+     * Create Log Replication Runtime for a session, if not already created.
      */
-    public void start(ClusterDescriptor remoteCluster, LogReplicationSession session,
-                      LogReplicationContext replicationContext) {
+    public void createRuntime(ClusterDescriptor remote, LogReplicationSession replicationSession,
+                              LogReplicationClientServerRouter router) {
         try {
-            // TODO (V2): we might think of unifying the info in ClusterDescriptor into session (all nodes host+port)
-            startLogReplicationRuntime(remoteCluster, session, replicationContext);
+            CorfuLogReplicationRuntime replicationRuntime;
+            if (!sessionRuntimeMap.containsKey(replicationSession)) {
+                log.info("Creating Log Replication Runtime for session {}", replicationSession);
+                LogReplicationRuntimeParameters parameters;
+                // parameters is null when the cluster is not the connection starter.
+                if (replicationSessionToRuntimeParams.isEmpty() || replicationSessionToRuntimeParams.get(replicationSession) == null) {
+                    parameters = createRuntimeParams(remote, replicationSession);
+                } else {
+                    parameters = replicationSessionToRuntimeParams.get(replicationSession);
+                }
+                replicationRuntime = new CorfuLogReplicationRuntime(parameters,
+                        metadataManager, upgradeManager, replicationSession, replicationContext, router);
+                sessionRuntimeMap.put(replicationSession, replicationRuntime);
+                router.addRuntimeFSM(replicationSession, replicationRuntime);
+            } else {
+                log.warn("Log Replication Runtime to remote session {}, already exists. Skipping init.",
+                        replicationSession);
+                return;
+            }
         } catch (Exception e) {
-            log.error("Failed to start log replication runtime for session={}", session);
+            log.error("Caught exception, stop log replication runtime to {}", replicationSession, e);
+            stopLogReplicationRuntime(replicationSession);
         }
     }
 
+    /**
+     * Start log replication.
+     * For the connection initiator cluster, this is called when the connection is established.
+     */
+    public void startLogReplicationRuntime(LogReplicationSession replicationSession) {
+        CorfuLogReplicationRuntime replicationRuntime;
+        synchronized (this) {
+            replicationRuntime = sessionRuntimeMap.get(replicationSession);
+        }
+        replicationRuntime.start();
+    }
```

Comment:
Why we acquire lock here only for the get method? I thought it didn't provide more synchronization than just ConcurrentHashMap?

Reply:
yes, this isn't needed now. removed.